perceval/backends/remo2.py

```diff
@@ -0,0 +1,318 @@
+from ..cache import Cache
+from ..errors import BackendError, CacheError, ParseError
+
+from ..utils import (DEFAULT_DATETIME,
+                     datetime_to_utc,
+                     str_to_datetime,
+                     urljoin)
+
+
+logger = logging.getLogger(__name__)
+
+MOZILLA_REPS_URL = "https://reps.mozilla.org"
+
+class ReMo(Backend):
+    """ReMo backend for Perceval.
+
+    This class retrieves the events from a
+    ReMo url. To initialize this class a
+    url could be provided. If not, https://reps.mozilla.org will be used.
+
+    It uses v2 API to get events, people and activities data.
+
+    :param url: ReMo url
+    :param cache: cache object to store raw data
+    :param origin: identifier of the repository; when `None` or an
+        empty string are given, it will be set to `url` value
+    """
+    version = '0.2.0'
+
+    def __init__(self, url=None, cache=None, origin=None):
+        if not url:
+            url = MOZILLA_REPS_URL
+        origin = origin if origin else url
+
+        super().__init__(origin, cache=cache)
+        self.url = url
+        self.client = ReMoClient(url)
+        self.__users = {}  # internal users cache
+
+    @metadata
+    def fetch(self, kind='events'):
+        """Fetch events from the ReMo url.
+
+        The method retrieves, from a ReMo url, the
+        events.
+
+        :returns: a generator of events
+        """
+
+        supported_kinds = ['activities', 'events', 'users']
+
+        if kind not in supported_kinds:
+            raise RuntimeError('ReMo perceval backend does not support %s', kind)
+
+        logger.info("Looking for events at url '%s'", self.url)
+
+        nitems = 0  # number of items processed
+        titems = 0  # number of items from API data
+
+        self._purge_cache_queue()
+
+        for raw_items in self.client.get_items(kind):
+            self._push_cache_queue(raw_items)
+            items_data = json.loads(raw_items)
+            titems = items_data['count']
+
+            items = items_data['results']
+            for item in items:
+                raw_item_details = self.client.call(item['_url'])
+                self._push_cache_queue(raw_item_details)
+                item_details = json.loads(raw_item_details)
+                yield item_details
+                nitems += 1
+
+                self._flush_cache_queue()
+
+        logger.info("Total number of events: %i (%i expected)", nitems, titems)
+
+    @metadata
+    def fetch_from_cache(self):
+        """Fetch the items from the cache.
+
+        :returns: a generator of items
+
+        :raises CacheError: raised when an error occurs accessing the
+            cache
+        """
+
+        logger.info("Retrieving cached ReMo items: '%s'", self.url)
+
+        if not self.cache:
+            raise CacheError(cause="cache instance was not provided")
+
+        cache_items = self.cache.retrieve()
+
+        nitems = 0
+
+        while True:
```

Comment:
This infinite loop will make `next(cache_items)` to raise an exception when the iterator is exhausted.